python/ql/lib/semmle/python/frameworks/Stdlib.qll

```diff
@@ -3939,6 +3939,176 @@ private module StdlibPrivate {
     }
   }
 
+  // ---------------------------------------------------------------------------
+  // Flow summaries for container methods
+  // ---------------------------------------------------------------------------
+  /** A flow summary for `copy`. */
+  class CopySummary extends SummarizedCallable {
+    CopySummary() { this = "collection.copy" }
+
+    override DataFlow::CallCfgNode getACall() {
+      result.(DataFlow::MethodCallNode).getMethodName() = "copy"
+    }
+
+    override DataFlow::ArgumentNode getACallback() { none() }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      exists(string content |
+        content = "ListElement"
+        or
+        content = "SetElement"
+        or
+        exists(DataFlow::TupleElementContent tc, int i | i = tc.getIndex() |
+          content = "TupleElement[" + i.toString() + "]"
+        )
+        or
+        exists(DataFlow::DictionaryElementContent dc, string key | key = dc.getKey() |
+          content = "DictionaryElement[" + key + "]"
+        )
+      |
+        input = "Argument[self]." + content and
+        output = "ReturnValue." + content and
+        preservesValue = true
+      )
+      or
+      input = "Argument[self]" and
+      output = "ReturnValue" and
+      preservesValue = false
```

Comment:
My thought were that the operation does not preserve the pointer, which also seems to be what your link says.
Our reasoning is generally on pointers, but if it helps us to deviate here we should certainly consider that.

Reply:
I realize that my link is a little stupid, since it's not a method... however, it will be recognized as a `MethodCallNode` in our analysis still :thinking: 

> Our reasoning is generally on pointers

I'm not sure. I don't think I've ever been part of a discussion where we talked about this aspect of our analysis.

I guess the more philosophical question is, if you copy the contents of a string to a new string located in different part of memory, is that data-flow or taint-flow? C++ has [strcpy as data-flow](https://github.com/github/codeql/blob/fe891746bf1d9c18ce488e178977f7fda2f5de3c/cpp/ql/lib/semmle/code/cpp/models/implementations/Strcpy.qll#L95-L97), so that's one datapoint.

I think it's hard to discuss this aspect, since it doesn't seem like we need it for anything? (or maybe we do still need it until we've sorted out our taint handling?)

I would still propose we make this perserve value, but also realize that we shouldn't spend forever discussing this small detail, so if you really don't want to do it, I'll probably live happily with this only being a taint-step.

Reply:
I should have put a reference, but could not immediately find it. @taus and I thought it through for type tracking and `LocalSourceNode`s at some point. But since this step is quite likely to be type preserving (also when we remove time), it is probably fine to make this a dataflow step..